### PR TITLE
Cql4bonnie qdm5 3 1060

### PR DESCRIPTION
**Pull request by c-monkey**
Converted Component from supporting just a scalar result to also date and code

JIRA: https://jira.mitre.org/browse/BONNIE-1060

TESTS: tbd

### CODE REVIEW

**Comment by mayerm94** [comment]
JIRA Test https://jira.mitre.org/browse/BONNIE-1077

File: app/assets/javascripts/views/patient_builder/data_criteria.js.coffee [hunk 1]
```diff
@@ -540,6 +567,12 @@ class Thorax.Views.EditCriteriaValueView extends Thorax.Views.BuilderChildView
 
       compare_collection = @values.findWhere(key: model_key)
 
+      # component code was put into another field to reuse the Thorax View
+      if (@model.get('type') == "CMP")
+        tmp = @model.get('code_list_id_cmp')
```
Comment (jbradl11) [inline app/assets/javascripts/views/patient_builder/data_criteria.js.coffee]: Do we need tmp? Could we just assign `@model.set code_list_id: @model.get('code_list_id_cmp')`?

File: lib/measures/patient_builder.rb [hunk 2]
```diff
@@ -251,7 +253,18 @@ def self.recursive_field_value_derivation(value, value_sets, name, entry)
         field_value["title"] = Measures::PatientBuilder.select_value_sets(field.code_list_id, value_sets)["display_name"] if field_value
       elsif field.type == "CMP"
         field_value["code"] = Measures::PatientBuilder.select_code(field.code_list_id, value_sets)
-        field_value["result"] =  {"scalar"=>value["value"], "units"=>value["unit"]}
+        cmp_type = value["type_cmp"]
+        field_value["result"] = {}
+        if cmp_type == "TS"
+          field_value["result"]["units"] = "UnixTime"
+          field_value["result"]["scalar"] = value['value'] #Time.at(value['value']/1000).utc.strftime('%Y%m%d%H%M%S')
```
Comment (jbradl11) [inline lib/measures/patient_builder.rb]: Do we need this comment?

File: app/assets/javascripts/views/patient_builder/data_criteria.js.coffee [hunk 3]
```diff
@@ -540,6 +567,12 @@ class Thorax.Views.EditCriteriaValueView extends Thorax.Views.BuilderChildView
...
+        @model.set code_list_id_cmp: @model.get('code_list_id')
+        @model.set code_list_id: tmp
```
Comment (jbradl11) [inline app/assets/javascripts/views/patient_builder/data_criteria.js.coffee]: Actually it seems like the variables are switched here where the `@model.set code_list_id:` should most likely be set to `@model.get('code_list_id')` and `@model.set code_list_id_cmp: @model.get('code_list_id')`

File: app/assets/javascripts/views/patient_builder/data_criteria.js.coffee (hunk 1 above)
Comment (c-monkey) [reply app/assets/javascripts/views/patient_builder/data_criteria.js.coffee]: I have updated the comment to remove your confusion. We need to swap the 2 values.

File: lib/measures/patient_builder.rb (hunk 2 above)
Comment (c-monkey) [reply lib/measures/patient_builder.rb]: Thanks, fixed.

File: app/assets/javascripts/views/patient_builder/data_criteria.js.coffee (hunk 3 above)
Comment (c-monkey) [reply app/assets/javascripts/views/patient_builder/data_criteria.js.coffee]: see above

**Comment by jbradl11** [comment]
Frontend tests are failing. 

  1) PatientBuilderView CQL laboratory test performed should have custom view for components
     Failure/Error: Expected ({ length: 0, prevObject: ({ 0: HTMLNode, length: 1 }), context: undefined, selector: 'label[for=value]' }) to exist.

  2) PatientBuilderView CQL laboratory test performed should have custom view for components
     Failure/Error: TypeError: undefined is not an object (evaluating 'editFieldValueView.$('label[for=value]')[0].innerHTML') in http://127.0.0.1:58815/assets/views/patient_builder_spec.js?body=1 (line 447)

**Review by jbradl11 — APPROVED** [review]
Jira test and unit tests pass. UI changes made. Looks good.

**Review by holmesie — APPROVED** [review]

**Pull request merged**